flink-runtime/src/test/java/org/apache/flink/runtime/io/network/api/writer/RecordWriterTest.java

```diff
@@ -247,136 +388,121 @@ public void testBroadcastEventMixedRecords() throws Exception {
-		assertEquals(2, queues[0].size()); // 1 buffer + 1 event
-		assertTrue(parseBuffer(queues[0].remove(), 0).isBuffer());
-		assertEquals(3, queues[1].size()); // 2 buffers + 1 event
-		assertTrue(parseBuffer(queues[1].remove(), 1).isBuffer());
-		assertTrue(parseBuffer(queues[1].remove(), 1).isBuffer());
-		assertEquals(2, queues[2].size()); // 1 buffer + 1 event
-		assertTrue(parseBuffer(queues[2].remove(), 2).isBuffer());
-		assertEquals(1, queues[3].size()); // 0 buffers + 1 event
+			for (int i = 0; i < numberOfChannels; i++) {
+				assertEquals(4, queues[i].size()); // 3 buffer + 1 event
 
-		// every queue's last element should be the event
-		for (int i = 0; i < numberOfChannels; i++) {
-			boe = parseBuffer(queues[i].remove(), i);
-			assertTrue(boe.isEvent());
-			assertEquals(barrier, boe.getEvent());
+				for (int j = 0; j < 3; j++) {
+					assertTrue(parseBuffer(queues[i].remove(), 0).isBuffer());
+				}
+
+				BufferOrEvent boe = parseBuffer(queues[i].remove(), i);
+				assertTrue(boe.isEvent());
+				assertEquals(barrier, boe.getEvent());
+			}
+		} else {
+			assertEquals(4, bufferProvider.getNumberOfCreatedBuffers());
+
+			assertEquals(2, queues[0].size()); // 1 buffer + 1 event
+			assertTrue(parseBuffer(queues[0].remove(), 0).isBuffer());
+			assertEquals(3, queues[1].size()); // 2 buffers + 1 event
+			assertTrue(parseBuffer(queues[1].remove(), 1).isBuffer());
+			assertTrue(parseBuffer(queues[1].remove(), 1).isBuffer());
+			assertEquals(2, queues[2].size()); // 1 buffer + 1 event
+			assertTrue(parseBuffer(queues[2].remove(), 2).isBuffer());
+			assertEquals(1, queues[3].size()); // 0 buffers + 1 event
+
+			// every queue's last element should be the event
+			for (int i = 0; i < numberOfChannels; i++) {
+				BufferOrEvent boe = parseBuffer(queues[i].remove(), i);
+				assertTrue(boe.isEvent());
+				assertEquals(barrier, boe.getEvent());
+			}
 		}
 	}
 
-	/**
-	 * Tests that event buffers are properly recycled when broadcasting events
-	 * to multiple channels.
-	 */
-	@Test
-	public void testBroadcastEventBufferReferenceCounting() throws Exception {
-
+	private void broadcastBufferOrEventReferenceCounting(boolean isBroadcastWriter, boolean isBroadcastEvent) throws Exception {
 		@SuppressWarnings("unchecked")
 		ArrayDeque<BufferConsumer>[] queues = new ArrayDeque[] { new ArrayDeque(), new ArrayDeque() };
 
-		ResultPartitionWriter partition =
-			new CollectingPartitionWriter(queues, new TestPooledBufferProvider(Integer.MAX_VALUE));
-		RecordWriter<?> writer = new RecordWriter<>(partition);
+		ResultPartitionWriter partition = new CollectingPartitionWriter(queues, new TestPooledBufferProvider(Integer.MAX_VALUE));
+		RecordWriter<IntValue> writer = isBroadcastWriter ?
+			RecordWriter.createRecordWriter(partition, new OutputEmitter(ShipStrategyType.BROADCAST, 0), "test") :
+			RecordWriter.createRecordWriter(partition);
 
-		writer.broadcastEvent(EndOfPartitionEvent.INSTANCE);
+		if (isBroadcastEvent) {
```

Comment:
+1 for the code deduplication, but could you do it in separate commit (preferably before the actual change)? Now it's confusing which changes are functional and which are caused by deduplication/refactoring.

Reply:
Yes, it would confuse you after I refactor the whole tests.
I could keep the previous tests unchanged, but if my new tests reuse some common codes from previous tests. Do you mind I extract the common parts together?  For the old tests not related to my new tests, I would keep them unchanged.

Reply:
I'm not saying that you should revert this code deduplication (definitely keep it!), but just extract it to a separate commit, to not mix refactor and functional changes.